src/v/cloud_storage/partition_recovery_manager.cc

```diff
@@ -390,10 +357,25 @@ partition_downloader::download_log_with_capped_size(
         .max_offset = model::offset::min(),
       }};
 
-    co_await ss::max_concurrent_for_each(
-      staged_downloads,
-      max_concurrency,
-      [this, &dlpart, &dloffsets](const segment_meta& s) -> ss::future<> {
+    // process from most recent (last one)
+    auto offset_segment_it = offset_map.rbegin();
+    auto offset_end = offset_map.rend();
+    size_t total_size = 0;
+    auto data_found = false;
+
+    co_await ss::do_until(
+      [&] {
+          return offset_segment_it == offset_end
+                 || // no more segments to process
+                    // or processing this segment would exceed max_size
+                    // and some proper data is ready
+                 (data_found
+                  && (total_size + offset_segment_it->second.size_bytes >= max_size));
+      },
+      [&]() -> ss::future<> {
+          auto [_, s] = *offset_segment_it;
+          ++offset_segment_it;
```

Comment:
@jcsp @andijcr this pattern here--where an iterator to a non-local container is held across scheduling points--is dangerous because all _other_ code paths that might asynchronously modify the container need to be safe with respect to the iterator that is held.

is this case safe? this is the sort of pattern that is likely to get flagged when looking into in segfaults and undefined behavior, so it's worth writing a comment to avoid that investigation cost.

Reply:
it should be safe here, offset_map is local to the parent function and the code path is linear https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/partition_recovery_manager.cc#L259
I'll write a comment, since this is not the way to go for the next iteration where a parallel or batch approach to downloads could be pursued

Reply:
a comment would be great, thanks. it's often more "future proof" to ignore the optimization of holding the iterator, and simply re-querying the container after every loop.